### ISSUE TITLE
recheck if files got changed since last update

### ISSUE THREAD

**Issue by tbeitter** [Sonerezh/sonerezh#127]
It seems that once imported track data will not be updated later on, even if the file was changed on the file system.

1. import your database
2. check track tags (f.e. album title)
3. change album title using a tagger of your choice
4. udpate the database
5. recheck album title in sonerezh, no change

Sonerezh should check if the file was updated since it got imported to the database. Right now there is no way to fix tag errors after the initial import, as sonerezh will not update tag details. A full database reset and import lasts too long and should not be a daily task.

Best,
Til

**Comment by lGuillaume124** [comment]
Indeed, this update operation will not be necessary when you will be able to edit your songs directly from Sonerezh ;)
This major (and must have) feature is planned for the 1.1.0 release.

**Comment by tbeitter** [comment]
fine. Even with the ability to edit files in sonerezh it would be nice if changed files would be recognised. Sometimes user run scripts or other crawlers to automatically tag hundreds of files this as it's faster then manually tagging it yourself.

**Comment by MightyCreak** [comment]
I agree with @tbeitter, I'm used to EasyTag so unless tag edition is as good, I probably won't use the in-app editor (not that it's not important, I'm sure other users will like it).

And even if I were to use the in-app editor, it won't solve all the possible scenarios:
* I changed the ID3 tags with another editor and re-uploaded it on my server
* I re-ripped my album with a better (or worse) quality
* I deleted an album on my server

Cheers!

**Comment by MightyCreak** [comment]
It's just my opinion, but I think that having a better updater is more important than editing my library directly in the web-app.

Right now we only have this rule:
* Update each files in the HDD that is not referenced in the DB

I think we should add these simple rules also for each files from the DB:
* Remove the song in the DB if the file behind `source_path` no longer exist
* Update the song in the DB if the modification date of `source_path` is different than the date in the `modified` field

That way, all the files in the HDD will be in sync with the songs in the DB. We would just have to click on "Update the database" ;)

**Comment by ThibautMarty** [comment]
+1
Even if the 1.1.1 is released, I think this is not yet fixed, right ?

In my situation, the Sonerezh's music directory is the download dir for torrents. When a torrent is incomplete, its file can exists though invalid.
On import, I get warnings:
`[music file] Metadata are unreadable or empty. Trying to import anyway...`
and the invalid files are imported. Then, there is no way to update them.

Like @MightyCreak, the tag edition will not solve this issue. And my only option is to reset the database :/

(Thanks for your work. It looks like the player I've always wanted (to build))

Thibaut

**Comment by maxengel** [comment]
I agree that it is more important to have existing tracks updated than a metadata editor within the app.

**Comment by MightyCreak** [comment]
I've found a workaround: delete the songs in phpMyAdmin, like that:

```sql
DELETE FROM `songs` WHERE `songs`.`album` = "your album title"
```

Then you can re-update your song database in Sonerezh.

It's better than reset the database because:
* it's faster
* you don't lose your playlists

Be careful though, it's been tested in v1.1.3, and there are no dependencies between the SQL tables in this version. At some point there might be some in future releases in order to have better performances (probably more 1.2 I guess).

**Comment by Amund** [comment]
Have solved this on a comics reader of mine, seems pretty simple and fast to me. Here is a stripped down version of that code, to adapt and optimize, where replace stand for inserting new elements and **updating** elements which have changed.

```php
function rglob( $pattern, $flags = 0 ) {
	$files = glob( $pattern, $flags );
	foreach( glob( dirname( $pattern ).'/*', GLOB_ONLYDIR|GLOB_NOSORT ) as $dir ) {
		$files = array_merge( $files, rglob( $dir.'/'.basename( $pattern ), $flags ) );
	}
	return $files;
}

function reindex( $pdo, $path_bd ) {
	// files on disk
	$list_disk = rglob( $path_bd.'/*.cbz' );
	sort( $list_disk );
	$list_disk = array_flip( $list_disk );
	foreach( $list_disk as $k=>$v )
		$list_disk[$k] = filemtime( $k );

	// files in db
	$list_db = array();
	$stmt = $pdo->query( 'SELECT path, mtime FROM bd_album' );
	while( $r = $stmt->fetch( PDO::FETCH_ASSOC ) )
		$list_db[$r['path']] = $r['mtime'];
	ksort( $list_db );

	// to delete
	$delete = array_diff_key( $list_db, $list_disk );
	if( count( $delete ) > 0 ) {
		foreach( $delete as $k=>$v ) {
			// ...delete element
		}
	}

	// to replace
	$replace = array_diff_assoc( $list_disk, $list_db );
	if( count( $replace ) > 0 ) {
		foreach( $replace as $k=>$v ) {
			// ...replace element
		}
	}
}
```

**Comment by gs11** [comment]
Fix available in PR #293

**Comment by MightyCreak** [comment]
@gs11 I think your PR fixes #129, but not this issue here.

**Comment by gs11** [comment]
Oops, removed my comment @MightyCreak

**Comment by gs11** [comment]
@MightyCreak I like the modification date check idea you mentioned in #310. Checking the modification date is fast enough without having to re-read the file to see if any tags have changed.

**Comment by gs11** [comment]
I did a fix for this in https://github.com/gs11/sonerezh/commit/26e3b3d1d915f0460278b12313e9cf672357aa95

The update process now checks for files that have a system modified time later than the modified column in the database. If so it re-parses the file and updates the existing record.

**Comment by gs11** [comment]
Fixed in v1.2.0, please close the issue.
